[NVIDIA XLA]Replace broadcast of trivial matrix bias by optional custom-call for CUDA >= 12

wenscarl opened the pull request:
Replace broadcast of trivial matrix bias by a scalar under CUDA > 12. Nearly all F8 unittests will get impacted but only 1 unittest is included at the moment for the purpose of demonstration. 

Broadcasting trivial matrix bias takes up about 10% time of the actually matmul. 
![image](https://user-images.githubusercontent.com/25590028/234122329-3ae6040e-c940-4131-95ce-0644411a41cd.png)

wenscarl committed:
Broadcast -> scalar 0

pjannaty commented:
@reedwm this seems to be of high priority given @wenscarl is reporting the perf hit is ~ 10% of the actual matmul call.

pjannaty commented:
Seems like all the unit tests will have to change to accommodate this change. Wenscarl has only implemented a single unit test to show the shape of things to come and is seeking feedback whether the new paradigm works. @philipphack will also give feedback.

wenscarl committed:
Fix typo

reedwm commented on tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -834,38 +836,66 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
+      PaddingConfig padding_config;
+      c_padded = !ShapeUtil::Equal(pad_shape(padding_config, instr->shape()),
+                                  instr->shape());
+      c = instr->AddInstruction(
+          HloInstruction::CreateConstant(LiteralUtil::Zero(c_type)));
```
Why create a `c` at all, if it's just going to be replaced by an empty buffer in the thunk? We don't have to pass `c` to the custom call.

wenscarl replied:
Yes, but c has to be something, in this case I made it a scalar, to get passed into HloInstruction::CreateCustomCall().

philipphack replied:
We could create the Custom Call without c when beta is zero and CUDA is version 12 or higher.

wenscarl replied:
Yes, it's possible, but that requires lots of plumbing work and hence generating quite several compile conditional flags. A scalar placeholder operand is used here for convenience.

reedwm replied:
I would much prefer creating the custom call without C when beta is zero. Otherwise it's inconsistent with the non-FP8 case and it's confusing how the custom call has an unused placeholder for C that is unused.

Having a lot of `CUDA_VERSION >= 12000` checks would be annoying. Is it possible to set beta to 1 for CUDA 11 when creating a zero bias, in which case we wouldn't need such checks and instead could assume we have a matrix bias iff beta is nonzero. Alternatively, perhaps we should only support this pass for CUDA 12 and above.

wenscarl replied:
If to implementing creating custom call without C to make it consistent with the non-FP8 , the operands[2] can be matrix or vector bias, check [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/translate/mhlo_to_lhlo_with_xla/mhlo_to_lhlo_with_xla.cc;l=971), which is what non-FP8 case does. But this plan was rejected in that PR. Check earlier discussion [here](https://github.com/tensorflow/tensorflow/pull/60020#discussion_r1145622210).

wenscarl commented on tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -834,38 +836,66 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
-    pad_operand(c);
+    bool c_padded = false;
+    absl::Span<const int64_t> c_dim;
+#if CUDA_VERSION > 12000
```
I saw [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc;l=728?q=12000&ss=tensorflow%2Ftensorflow) is > 12000. So follow that.

wenscarl commented on tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -834,38 +836,66 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
-      PaddingConfig padding_config;
-      Shape padded_shape = x->shape();
-      for (int i = 0; i < x->shape().rank(); ++i) {
+    auto pad_shape = [&batch_dims](PaddingConfig &padding_config,
```
Yes. Isolate `pad_shape` to be used in determining if c is padded.

wenscarl commented on tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -834,38 +836,66 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
+    if (c == nullptr) {
+      c_dim = instr->shape().dimensions();
+      PaddingConfig padding_config;
+      c_padded = !ShapeUtil::Equal(pad_shape(padding_config, instr->shape()),
```
Yes, but c has to be something, in this case I made it a scalar, to get passed into `HloInstruction::CreateCustomCall()`.

wenscarl committed:
Fix typo.

wenscarl committed:
Update CUDA_VERSION guard

wenscarl committed:
Update CUDA_VERSION guard

wenscarl committed:
Remove C_dim proxy.

wenscarl committed:
Remove dummy parameters

wenscarl committed:
Update unittests

philipphack commented:
Version 11 doesn't support non-zero beta. I think that's why we added the version guard [here](https://github.com/tensorflow/tensorflow/blame/0e51c74cd0fa8bef5d9cdbea76f3387a1baae01b/tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc#L728).

wenscarl committed:
[WIP] Remove placeholder C.

wenscarl committed:
Pass unittests.

wenscarl committed:
Remove CUDA_VERSION guards

wenscarl commented:
Changes are made with CUDA_VERSION >= 12000 guards removed.

reedwm commented:
Can you sync with the master branch to get changes from https://github.com/tensorflow/tensorflow/pull/60502?

wenscarl commented:
> Can you sync with the master branch to get changes from #60502?

Rebased to current master.

reedwm commented on tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -726,7 +726,6 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     // fusion of the scaling and conversion of D into the Custom Call. Fusing
     // a matrix bias is only supported with CUDA 12 and above.
     HloInstruction *c = nullptr;
-
```
Revert whitespace changes

reedwm commented on tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:
```diff
@@ -867,16 +857,41 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     };
+
+    // Get the possible padded shape.
+    auto pad_shape = [&batch_dims](const Shape old_shape) {
```
Use `pad_shape` it in `pad_operand` to avoid the duplicated code which calls `RoundUpTo`.

Also it might be worth making `pad_shape` and `pad_operand` top-level functions instead of lambda functions.

reedwm commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1164,8 +1164,12 @@ Status IrEmitterUnnested::EmitCublasLtMatmulThunkF8(mlir::Operation* op) {
                       GetAllocationSlice(matmul.getB()));
-  TF_ASSIGN_OR_RETURN(BufferAllocation::Slice c,
-                      GetAllocationSlice(matmul.getC()));
+  BufferAllocation::Slice c;
+
+  if (matmul.getBetaAttr().getValueAsDouble() != 0.0) {
+    TF_ASSIGN_OR_RETURN(c, GetAllocationSlice(matmul.getC()));
+  }
+
```
The non-FP8 case doesn't have this logic. It unconditionally gets C, which IIUC is simply the output if there is no matrix bias. For consistency with the non-FP8 case, can we also uncondtionally get C here? This would also simply some logic in cublas_lt_matmul_think.cc and matmul_utils.h.

reedwm reviewed (commented):
Since @wenscarl is currently OOO, I'm approving and will address all comments when merging

reedwm committed:
Merge branch 'master' into f8_matrix_bias_no_broadcast

tdanyluk commented:
As part of the OpenXLA PR rotation:
I'll re-trigger the presubmits as soon as I get permissions.

reedwm commented:
Since the PR author is OOO, I was going to fix some issues when merging. But it's too hard to test on H100s when merging, so instead I'll commit directly to this PR's branch. So let's not merge this PR yet.

reedwm committed:
Merge branch 'master' into f8_matrix_bias_no_broadcast

reedwm commented:
@philipphack in the FP8 case, when there is no matrix bias, this PR currently sets C to nullptr. But in the FP16 case, when there is no matrix bias, instead C is set to D

I would prefer that the FP16 and FP8 cases be unified. Is there a fundamental reason why it's important for C to be nullptr in the FP8 case but not in the non-FP8 case? If there is no difference in cublas LT between FP8 and FP16 w.r.t. whether C should be nullptr or equal to D, I will set C to D in this PR. (Since @wenscarl is OOO I am addressing my own comments to this PR before merging it)

reedwm committed:
Address my own comments.

reedwm committed:
Fix failing test.

Fix test //tensorflow/compiler/xla/translate/mhlo_to_lhlo_with_xla/tests:hlo_text_to_lhlo_no_opt.hlotxt.test.
